src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java

```diff
@@ -0,0 +1,288 @@
+                Arrays.copyOf(accessModifiers, accessModifiers.length);
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Count record components in this record definition
+        final ComponentCounter counter = countComponents(ast);
+
+        // Check counter and report violations
+        if (matchAccessModifiers(counter.getAccessModifierOption())
+            && counter.getTotal() > max) {
+            log(ast, MSG_KEY, counter.getTotal(), max);
+        }
+    }
+
+    /**
+     * Determine the visibility modifier, then iterate through the record component list
+     * and increment the counter.
+     *
+     * @param ast the ast to check
+     * @return instance of component counter with count of components and
+     *     access modifier of record definition.
+     */
+    private static ComponentCounter countComponents(DetailAST ast) {
+        final ComponentCounter counter = new ComponentCounter();
+        final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+
+        counter.setAccessModifierOption(getAccessModifier(ast));
+
+        TokenUtil.forEachChild(recordComponents,
+                TokenTypes.RECORD_COMPONENT_DEF, node -> counter.increment());
+
+        return counter;
+    }
+
+    /**
+     * Returns the access modifier of the record definition at the specified AST. If
+     * the record definition is in an interface or annotation block,
+     * the access modifier is assumed to be public.
+     *
+     * @param ast the token of the method/constructor.
+     * @return the access modifier of the method/constructor.
+     */
+    private static AccessModifierOption getAccessModifier(final DetailAST ast) {
+        final DetailAST modsToken = ast.findFirstToken(TokenTypes.MODIFIERS);
+        return CheckUtil.getAccessModifierFromModifiersToken(modsToken);
+    }
+
+    /**
+     * Checks whether a record definition has the correct access modifier to be checked.
+     *
+     * @param accessModifier the access modifier of the record definition.
+     * @return whether the record definition matches the expected access modifier.
+     */
+    private boolean matchAccessModifiers(final AccessModifierOption accessModifier) {
+        return Arrays.stream(accessModifiers)
+                .anyMatch(modifier -> modifier == accessModifier);
+    }
+
+    /**
+     * Counter class used to track the access modifier of the record definition and the number of
+     * components in the record component list.
+     */
+    private static class ComponentCounter {
```

Comment:
Please remove this class, to just count you need single variable.

Reply:
Removed, thank you.